pkg/controllers/networkpolicy/policy_combiner.go

```diff
@@ -0,0 +1,61 @@
+package networkpolicy
+
+import (
+	log "github.com/sirupsen/logrus"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sort"
+)
+
+func combinePolicies(nps []*networkingv1.NetworkPolicy) *networkingv1.NetworkPolicy {
+	// policies should be combined in alphabetical order
+
+	sort.Slice(nps, func(i, j int) bool {
+		return nps[i].Name < nps[j].Name
+	})
+
+	base := &networkingv1.NetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nps[0].Labels[dynamicLabel],
+			Namespace: nps[0].Namespace,
+		},
+		Spec: networkingv1.NetworkPolicySpec{
+			PodSelector: nps[0].Spec.PodSelector,
+			PolicyTypes: nps[0].Spec.PolicyTypes,
+		},
+	}
+
+	baseSelector := base.Spec.PodSelector.String()
+	for _, np := range nps {
+		if np.Namespace != base.Namespace {
+			log.Warn("Namespaces aren't equal for all network policies labelled with monzo.com/dynamic-networkpolicy=", base.Name, ". skipping ", np.Name)
+			continue
+		}
+
+		if np.Spec.PodSelector.String() != baseSelector {
+			log.Warn("Selectors aren't equal for all network policies labelled with monzo.com/dynamic-networkpolicy=", base.Name, ". skipping ", np.Name)
+			continue
+		}
+
+		base.Spec.Ingress = append(base.Spec.Ingress, np.Spec.Ingress...)
+		base.Spec.Egress = append(base.Spec.Egress, np.Spec.Egress...)
+	}
+
+	return base
+}
+
+func interfacesToNetPols(is []interface{}) []*networkingv1.NetworkPolicy {
+	nps := make([]*networkingv1.NetworkPolicy, 0, len(is))
+	for _, i := range is {
+		nps = append(nps, i.(*networkingv1.NetworkPolicy))
+	}
+	return nps
+}
+
+func combinerIndexFunc(obj interface{}) ([]string, error) {
+	if ann, ok := obj.(*networkingv1.NetworkPolicy).Labels[dynamicLabel]; ok {
```

Comment:
This looks strange, I didn't know that would work to cast it and then get something out of that.

Reply:
It does work and it panics if obj is a different type, but we have some guarantees from the informer that this isn't possible

Reply:
As we discussed, I mistakenly thought this `ann, ok` was the result of a safe type cast, but in fact it's the result of the Labels[dynamicLabel] map access.